azkaban-common/src/main/java/azkaban/executor/ExecutionFlowDao.java

```diff
@@ -338,13 +339,64 @@ public int selectAndUpdateExecution(final int executorId, final boolean isActive
     }
   }
 
+  public int selectAndUpdateExecutionWithLocking(final int executorId, final boolean isActive)
+      throws ExecutorManagerException {
+    final String UPDATE_EXECUTION = "UPDATE execution_flows SET executor_id = ?, update_time = ? "
+        + "where exec_id = ?";
+    final String selectExecutionForUpdate = isActive ?
+        SelectFromExecutionFlows.SELECT_EXECUTION_FOR_UPDATE_ACTIVE :
+        SelectFromExecutionFlows.SELECT_EXECUTION_FOR_UPDATE_INACTIVE;
+
+    final SQLTransaction<Integer> selectAndUpdateExecution = transOperator -> {
+      int execId = -1;
+      final boolean hasLocked = transOperator.query(PollingLockResultHandler.GET_POLLING_LOCK, new PollingLockResultHandler());
+      logger.info("ExecutionFlow polling lock value: " + hasLocked + " for executorId: " + executorId);
+      if (hasLocked) {
+        try {
+          final List<Integer> execIds = transOperator.query(selectExecutionForUpdate, new SelectFromExecutionFlows(), executorId);
+          if (CollectionUtils.isNotEmpty(execIds)) {
+            execId = execIds.get(0);
+            transOperator.update(UPDATE_EXECUTION, executorId, System.currentTimeMillis(), execId);
+          }
+        } finally {
+          transOperator.query(PollingLockResultHandler.RELEASE_POLLING_LOCK, new PollingLockResultHandler());
+          logger.info("Released polling lock for executorId: " + executorId);
+        }
+      }
+      return execId;
+    };
+
+    try {
+      return this.dbOperator.transaction(selectAndUpdateExecution);
+    } catch (final SQLException e) {
+      throw new ExecutorManagerException("Error selecting and updating execution with executor "
+          + executorId, e);
+    }
+  }
+
+  private static class PollingLockResultHandler implements ResultSetHandler<Boolean> {
```

Comment:
Can we make this class generic and move outside as it is trying to get lock and this can be leveraged in other place wherever required.